workflow/libero_flow_poc/libero_flow/decider.py

```diff
@@ -0,0 +1,196 @@
+fh = logging.FileHandler(f'{__name__}.log')
+logger.addHandler(fh)
+
+
+def send_decision_message(decision: Dict):
+    """create and send decision task message.
+
+    :param decision: dict
+    :return:
+    """
+    with get_channel() as channel:
+        message = get_base_message()
+        message['aggregate']['service'] = 'flow-decider'
+        message['aggregate']['name'] = 'workflow-decision'
+        message['type'] = 'decision-task-result'
+        message['data'] = decision
+
+        channel.basic_publish(exchange=DECISION_RESULT_EXCHANGE,
+                              routing_key=DECISION_RESULT_QUEUE,
+                              body=json.dumps(message),
+                              properties=pika.BasicProperties(delivery_mode=DELIVERY_MODE_PERSISTENT))
+        print(f'[x] Schedule decision sent: {message}')
+
+
+def decide(workflow: Dict) -> Dict:
+    """Look at the current workflow state and decide on the next action.
+
+    example `workflow` value:
+
+    {
+        "instance_id": "8eed4f02-4d3c-4fb2-89f8-1507374ae541",
+        "name": "FooBarWorkflow",
+        "status": "Pending",
+        "created": "2018-04-11T12:23:34.236207Z",
+        "start_timestamp": null,
+        "end_timestamp": null,
+        "config": {
+            "timeout": 300
+        },
+        "input_data": {
+            "timeout": 12355
+        },
+        "activities": [
+            {
+                "instance_id": "c058cd3f-93c4-4d2d-a0f4-d6eae5a3e6e2",
+                "name": "PingWorker",
+                "independent": false,
+                "required": true,
+                "status": "Pending",
+                "config": {
+                    "foo": "bar"
+                },
+                "workflow": "8eed4f02-4d3c-4fb2-89f8-1507374ae541"
+            },
+            {
+                "instance_id": "e05678d0-20b0-4a75-8452-4be151c71461",
+                "name": "SumValues",
+                "independent": false,
+                "required": true,
+                "status": "Pending",
+                "config": {
+                    "foo": "bar"
+                },
+                "workflow": "8eed4f02-4d3c-4fb2-89f8-1507374ae541"
+            }
+        ]
+    }
+
+    :param workflow:
+    :return: dict
+    """
+    decision = {
+        'workflow_id': workflow['instance_id'],
+        'decision': ''
+    }
+
+    # TODO replace temp logic tree with more extendable methods?
+    # check current workflow state
+    if workflow['status'] == FINISHED:
+        decision['decision'] = 'do-nothing'
+
+    elif workflow['status'] == CANCELLED:
+        decision['decision'] = 'do-nothing'
+
+    elif workflow['status'] == PENDING:
+        decision['decision'] = 'start-workflow'
+
+    elif workflow['status'] == IN_PROGRESS:
+        # do some real work
+        activities_to_schedule = []
+        decision['decision'] = 'schedule-activities'
+
+        for activity in workflow['activities']:
```

Comment:
why are we going through multiple activities here? I thought only one would be started at a time... unless this is to keep as much parallelism as possible (like branching 3 activities to be run simultaneously)

Reply:
Yes, if a workflow has multiple independent activities then they will be started at the same time as they do not depend on other activity chaining or values.